library/Backends/llvm/src/Juvix/Backends/LLVM/Pass/Types.hs

```diff
@@ -0,0 +1,132 @@
+import qualified Juvix.Core.Erased.Ann as ErasedAnn
+import Juvix.Library
+import qualified Juvix.Library.NameSymbol as NameSymbol
+import qualified Juvix.Library.Usage as Usage
+
+--------------------------------------------------------------------------------
+-- Closure Capture Forms
+--------------------------------------------------------------------------------
+
+data Capture = Capture
+  { -- | @location@ represents where in the environment do we
+    -- capture the value from.
+    location :: CaptureFrom,
+    -- | @slot@ represents the new slot that represents this
+    -- captured value, for proper GEP generation.
+    slot :: ArraySlot,
+    -- | @capType@ represents the capture type of the captured
+    -- argument
+    capType :: ErasedAnn.Type Prim.PrimTy
+  }
+  deriving (Show)
+
+----------------------------------------
+-- Slot Location Storage Types
+----------------------------------------
+
+-- | @ArraySlot@ represents the slot in the array the node now is in
+-- along with metadata associated with the slot.
+data ArraySlot = Slot
+  { -- | @previousName@ represents the old name in the source code,
+    -- recorded for debugging purposes.
+    previousName :: NameSymbol.T,
+    -- | @newIndex@ represents the new slot location into the array.
+    newIndex :: Index
+  }
+  deriving (Show)
+
+-- | @Index@ is the index into the array
+newtype Index = Index {num :: Natural} deriving (Show)
+
+----------------------------------------
+-- Storage Location and Indexing Types
+----------------------------------------
+
+data CaptureFrom
+  = -- | @FromAmbiantEnv@ represents the values that we capture having a
+    -- proper name rather then being offsets into the environment.
+    FromAmbiantEnv NameSymbol.T
+  | -- | @FromClosureEnv@ represents that the value that we are trying
+    -- to store are offsets inside some environment, rather than a name
+    -- that we can directly bind.
+    FromClosureEnv IndexInto
+  deriving (Show)
+
+-- | @IndexInto@ represents the Index into an environment, along with
+-- which environment it originates from.
+data IndexInto = IndexInto
+  { index :: Index,
+    into :: FunctionEnvironment
+  }
+  deriving (Show)
+
+-- | @FunctionEnvironment@ represents if the index into an array is
+-- from the closure array/environment or the closure argument array/environment
+data FunctionEnvironment
+  = ClosureEnvironment
+  | -- Currently unused due to how it's impossible to distinguish
+    -- between functions and closures. We should in time compile with
+    -- this, and will be needed for currying
+    ArgumentEnvironemnt
+  deriving (Show)
+
+--------------------------------------------------------------------------------
+-- New Core Form we will Process Over
+--------------------------------------------------------------------------------
+
+data Annotated term = Ann
+  { usage :: Usage.T,
+    annTy :: ErasedAnn.Type Prim.PrimTy,
+    term :: term
+  }
+  deriving (Show)
```

Comment:
What's the reason for using this type instead of the existing
```
data AnnTerm primTy primVal = Ann
  { usage :: Usage.T,
    type' :: Type primTy,
    term :: Term primTy primVal
  }
  deriving (Show, Read, Eq, Generic)
```
from `Erased.Ann`? They look identical

Reply:
see `term` itself where the `Erased.Ann` version fixes it to the `Erased.Ann.Term` 

```haskell
term :: Term primTy primVal
```
Likewise I had to make a new `AnnTerm` which holds my own `Term`